MAINT: Const qualify UFunc inner loops

Kai-Striega opened the pull request:
This PR const qualifies the ``dimension`` and ``strides`` arguments of  ``PyUFuncGenericFunction``. Const qualified arguments make it simpler to reason about the behaviour of these ufuncs and prevents accidental mutation. As the const is now required this PR also const qualifies calls to ``PyUFuncGenericFunction`` inside the NumPy source code.

This closes #15252

Kai-Striega committed:
MAINT: Const qualify UFunc inner loops

This PR const qualifies the ``dimension`` and ``strides`` arguments of
``PyUFuncGenericFunction``. Const qualified arguments make it simpler
to reason about the behaviour of these ufuncs and prevents accidental
mutatation.

As the const is now required this PR also const qualifies calls to
``PyUFuncGenericFunction`` inside the NumPy source code.

Kai-Striega commented on numpy/__init__.pxd:
```diff
@@ -841,7 +841,7 @@ cdef inline char* _util_dtypestring(dtype descr, char* f, char* end, int* offset
 cdef extern from "numpy/ufuncobject.h":
 
-    ctypedef void (*PyUFuncGenericFunction) (char **, npy_intp *, npy_intp *, void *)
+    ctypedef void (*PyUFuncGenericFunction) (char **, const npy_intp *, const npy_intp *, void *)
```
Here and below I'm using west const (``(char **, const npy_intp *, const npy_intp *, void *)``). This is different to much of the existing east const code as cython doesn't seem to compile east const expressions. Should this be added as a note/fixme? And should this be added as a Cython issue?

eric-wieser replied:
As far as I can tell, we do not include any `const` in our cython code, so ok to omit this for now.

I would definitely be in favor of filing a feature request against cython to support east const :)

Kai-Striega replied:
> I would definitely be in favor of filing a feature request against cython to support east const :)

 It looks like there is an open issue already see [cython-1475](https://github.com/cython/cython/issues/1475).

> As far as I can tell, we do not include any const in our cython code, so ok to omit this for now.

Ok. I've removed the const qualifiers for now. It looks like Cython added const support in v0.18 with the latest release being 0.28 - I don't know the minimum NumPy requirement.

eric-wieser reviewed (commented):
Generally looks good. Please push any fixups as separate commits, to make subsequent review easier - we can squash at merge time.

eric-wieser commented:
Can you remove this now-incorrect comment?

https://github.com/numpy/numpy/blob/06a19d6a362af2508d1e11dce47eb2d18a419383/numpy/core/src/umath/ufunc_object.c#L3505-L3510

Kai-Striega committed:
REVERT: Remove const qualifiers from Cython code

Kai-Striega committed:
MAINT: Remove (now) incorrect comment and cast

Previously the `innerlooop` (a  PyUFuncGenericFunction) allowed
pointers to non-const qualified dimensions. PyUFuncGenericFunction now
only accepts const qualified dimensions making the cast, and explaining
comment, obselete.

Kai-Striega committed:
MAINT: Add release note

charris commented:
> I don't know the minimum NumPy requirement.

Cython 0.28.14 for Python 3.8 support.

eric-wieser commented:
Let's leave cython const correctness to a follow-up

seberg commented on doc/release/upcoming_changes/15355.c_api.rst:
```diff
@@ -0,0 +1,8 @@
+Const qualify UFunc inner loops
+-------------------------------
+`UFuncGenericFunction` now expects pointers to const ``dimension`` and
```
```suggestion
``UFuncGenericFunction`` now expects pointers to const ``dimension`` and
```
probably breaks release notes build otherwise. But its currently failing anyway, so cannot see (have to mop that up). In any case, looks good to me. Should merge this soon.

Kai-Striega committed:
Use double backticks for code snippet 15355.c_api.rst

Co-Authored-By: Sebastian Berg <sebastian@sipsolutions.net>

seberg commented:
Thanks @Kai-Striega and Eric. The cython issue is addressed, the release notes seem fine to me and the changes also look good, so putting this in.

rgommers commented:
I landed on this issue in a roundabout way. It looks like this inadvertently broke the ABI, which affected TensorFlow: https://github.com/tensorflow/tensorflow/issues/41902 (among others). 

Comment in current TensorFlow master branch:
```
    # NOTE: As numpy has releases that break semver guarantees and several other
    # deps depend on numpy without an upper bound, we must install numpy before
    # everything else.
    'numpy ~= 1.19.2',
```
The `~=` means `>= 1.19.2; < 1.20`.

The ABI break is a done deal now, I'm just commenting on it because it looks like we weren't even aware of it - is that right @seberg, @eric-wieser?

The version pinning thing is something we really should address with some guidance. What TensorFlow does is too restrictive and is going to cause problems if more packages start doing that. But the habit of not specifying any upper bound is indeed a real problem, we should get package to change that. I'll open a new issue for this.

eric-wieser commented:
I was completely unaware this was an ABI change. Does the tensorflow issue explain what the ABI issue is?

rgommers commented:
The one I linked to doesn't, but this may: https://github.com/tensorflow/tensorflow/pull/40728#issuecomment-671092945

eric-wieser commented:
My reading of this is that the reviewers were confused, and concluded ultimately that this was an API breakage (which we did know), not an ABI breakage.

seberg commented:
I don't know, but it sounds like a lot of confusion about "link time", but there is no link time with regard to the NumPy API (since we use an API function table instead).  If there was an actual ABI break, SciPy and a whole lot of things would have gone down in flames.  (How did SciPy fix this? Ignore the compile warnings?)

We were aware that this might force users to ignore warning or const qualify there loops. I did not realize that you might have to do both to get a warning free compile on either version of NumPy.  Or am I misunderstanding some C++ foo?

rgommers commented:
That discussion is indeed confusing. They seemed to agree on API break in the end. The trouble seems to be more that the impact was underestimated in gh-15252. You briefly said _"force users to add `const` to their loop signatures"_, but the rest of the discussion was only about compiler warnings, not that this would break already released packages. The issue doesn't even say "this will be an API break".

rgommers commented:
Maybe it's only a warning in C, but a hard error in C++?

seberg commented:
@rgommers do you have an example of an already released package breaking?  I don't understand how symbol names can matter considering the way our API is exported!

rgommers commented:
Never mind, I think that was more incorrect info. https://github.com/tensorflow/tensorflow/issues/41902 says that opting in to the new `pip` resolver, which got people a different `numpy` version, broke `tensorflow`. But reading the long linked Pip issue I don't think that's correct - it involved a source build breaking.

seberg commented:
OK, so the problem is that we were only aware that this will cause compiler warnings in C (which seems acceptable and even good). But, it actually causes C++ to barf, because C++ is much stricter about `const` (making this a real API break).

As is, I guess a code fix in `tensorflow` should not have been awfully hard. In some comments it sounds like the reasoning for not doing this (as in a forced cast) was fearing an actual ABI/API break :disappointed:; which simply won't happen.
(What will probably happen at some point is that we deprecate the `PyUFunc_RegisterLoopForDescr` function through Python warnings.)

In any case, even forcing `tensorflow` to modify code to compile successfully for both versions was obviously never intended here.  It would have been nice to know about the problem at a time when reverting might still have been helpful, we could even have had a 1.19.1 within days of the problem being noticed (and since it was not an ABI break, probably very few would have been hit).

(I don't know C++ well enough to know if there might be some other actual "warning only" solution for C++.)

eric-wieser commented:
> You briefly said "force users to add const to their loop signatures", but the rest of the discussion was only about compiler warnings, not that this would break already released packages.

Yes, I guess I forgot that already-released source (but not binary) packages are affected by API changes when users build them from source against the latest numpy.
